### DISCUSSION TITLE
Violation error when using set command

### DISCUSSION THREAD

**Discussion by Lukkyz** [prisma/prisma#18212] [Q&A]
Hi all,
I want to set files for a product with the set command, here's my schema :
```prisma
model Products {
  id                 Int                  @id @default(autoincrement())
  sku                String?              @unique(map: "sku") @db.VarChar(255)
  name               String?              @db.VarChar(255)
  price              Float?               @db.Float
  description        String?              @db.Text
  createdAt          DateTime             @default(now())
  updatedAt          DateTime             @default(now())
  files              Product_Files[]

  @@index([parent_product_id], map: "parent_product_id")
}
```

```prisma
model Files {
  id            Int             @id @default(autoincrement())
  name          String?         @db.VarChar(255)
  type          String?         @db.VarChar(255)
  url           String?         @db.VarChar(255)
  createdAt     DateTime        @default(now())
  updatedAt     DateTime        @default(now())
  Product_Files Product_Files[]
}
```

```prisma
model Product_Files {
  id         Int      @id @default(autoincrement())
  product_id Int
  file_id    Int
  createdAt  DateTime @default(now())
  updatedAt  DateTime @default(now())
  Products   Products @relation(fields: [product_id], references: [id], map: "Product_Files_ibfk_1")
  Files      Files    @relation(fields: [file_id], references: [id], map: "Product_Files_ibfk_2")

  @@index([file_id], map: "file_id")
  @@index([product_id], map: "product_id")
}
```

When I update a product, I want to delete old files, and add new files so i achieve this by doing that : 
```ts
prisma.products.update({
      where: {
        id,
      },
      data: {
        name,
        files: {
          deleteMany: {},
          createMany: {
            data: selectedFiles.map((file) => {
              return { file_id: file.id };
            }),
          },
        },
        sku,
        description,
        category_id: parseInt(category_id),
      },
    });
```
I discovered the set command for achieving the same, so I change the files object to :
```
files: {
     set: selectedFiles.map((file) => {
         return { id: file.id };
     }),
 },
```

But I get this error :

```
The change you are trying to make would violate the required relation 'Product_FilesToProducts' between the `Product_Files` and `Products` models.
code: 'P2014',
  clientVersion: '4.10.0',
  meta: {
    relation_name: 'Product_FilesToProducts',
    model_a_name: 'Product_Files',
    model_b_name: 'Products'
  },
  batchRequestIdx: undefined
```
Can you help me to figure what is the problem ?

**Comment by ludralph** [comment]
HI @Lukkyz 👋 

One possible reason for the error you're seeing could be because the Product_Files table has a required relationship with the Products table. This means that every Product_Files record must have a corresponding Products record.

When you use the set command to update the files field on the Products model, Prisma tries to update the records in the Product_Files table to match the new file IDs. However, if a file ID is removed from the files array but it is still associated with a Product_Files record, this violates the required relationship with the Products table and triggers the error.

To avoid this error while still using the set command, you can first remove the existing Product_Files records that are associated with the files that will be removed. You can then safely update the files field with the new set of file IDs.

Here's an example of how you could modify your code to use the set command and remove the old Product_Files records first:

```js
const res = await prisma.$transaction([
  prisma.product_Files.deleteMany({
    where: {
      product_id: 1,
      file_id: {
        notIn: selectedFiles.map((file) => file.id),
      },
    },
  }),
  prisma.products.update({
    where: {
      id: 1,
    },
    data: {
      name:"ABC",
      files: {
        set: selectedFiles.map((file) => ({ id: file.id })),
      },
       sku,
      description,
      category_id: parseInt(category_id),
    },
  }),
]);
```
The first query deletes any Product_Files records associated with files that are not in the new set of files. The second query updates the files field on the Products model with the new set of file IDs using the set command.

I hope this helps! Feel free to let me know if you have any more questions.

**Reply by ludralph** [reply]
Hey @Lukkyz,

do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by ludralph** [reply]
Hi @Lukkyz 

You may have missed my previous comments. Do let us know if you have any other questions or you can mark this discussion as answered.
